relayer/relayer.go

```diff
@@ -201,18 +208,32 @@ func NewRelayer(
 // On subscriber error, attempts to reconnect and errors if unable.
 // Exits if context is cancelled by another goroutine.
 func (r *Relayer) ProcessLogs(ctx context.Context) error {
+	doneCatchingUp := false
 	for {
 		select {
+		case catchUpResult := <-r.catchUpResultChan:
+			if !catchUpResult {
+				r.healthStatus.Store(false)
+				r.logger.Error(
+					"Failed to catch up on historical blocks. Exiting relayer goroutine.",
+					zap.String("originChainId", r.sourceBlockchainID.String()),
+				)
+				return fmt.Errorf("failed to catch up on historical blocks")
+			} else {
+				doneCatchingUp = true
+			}
 		case txLog := <-r.Subscriber.Logs():
+			// Relay the message to the destination chain. Continue on failure.
 			r.logger.Info(
 				"Handling Teleporter submit message log.",
 				zap.String("txId", hex.EncodeToString(txLog.SourceTxID)),
 				zap.String("originChainId", r.sourceBlockchainID.String()),
 				zap.String("sourceAddress", txLog.SourceAddress.String()),
 			)
 
-			// Relay the message to the destination chain. Continue on failure.
-			err := r.RelayMessage(&txLog)
+			// Messages are either catch-up messages, or live incoming messages.
+			// For live messages, we only write to the database if we're done catching up.
```

Comment:
nit: can you add a comment for why we can store processed height for catch up messages

Reply:
Sure thing, expanded this comment